src/NuGet.Clients/NuGet.SolutionRestoreManager/SolutionRestoreWorker.cs

```diff
@@ -486,21 +512,76 @@ await _solutionLoadedEvent.WaitAsync()
                             {
                                 if (isAllProjectsNominated)
                                 {
-                                    // if we've got all the nominations then continue with the auto restore
-                                    break;
+                                    if (isBulkRestoreCoordinationEnabled)
+                                    {
+                                        var projectReadyCheckMeasurement = Stopwatch.StartNew();
+                                        if (bulkRestoreCoordinationCheckStartTime == default)
+                                        {
+                                            bulkRestoreCoordinationCheckStartTime = DateTime.UtcNow;
+                                        }
+                                        projectsReadyCheckCount++;
+                                        // If we are about to start restore, we should run through all the projects to ensure there isn't a pending nomination.
+                                        var restoreProjectInfoSources = _solutionManager.Value.GetAllProjectRestoreInfoSources();
+                                        var allProjectsReady = true;
+                                        var bulkCheckTimeout = false;
+                                        for (int i = 0; i < restoreProjectInfoSources.Count && !bulkCheckTimeout; i++)
+                                        {
+                                            var restoreInfoSource = (IVsProjectRestoreInfoSource)restoreProjectInfoSources[i]; 
+                                            if (restoreInfoSource.HasPendingNomination)
+                                            {
+                                                allProjectsReady = false;
+                                                var timeoutTime = (bulkRestoreCoordinationCheckStartTime - DateTime.UtcNow).Value + BulkRestoreCoordinationTimeout;
```

Comment:
just to make sure if this change is intentional or below code is expected.

```cs
var timeoutTime = (DateTime.UtcNow - bulkRestoreCoordinationCheckStartTime).Value + BulkRestoreCoordinationTimeout;
```

Reply:
It looks `BulkRestoreCoordinationTimeout` ~ 5 min is max total allowed time for timeout. If we change to `(DateTime.UtcNow - bulkRestoreCoordinationCheckStartTime).Value + BulkRestoreCoordinationTimeout;` then it may result 13+ 5 = 18 min timeout. It's may not be the desired amount of time. With current logic it always keep it under 5 min.
Instead 

```Suggestion
TimeSpan timeoutTime = (bulkRestoreCoordinationCheckStartTime - DateTime.UtcNow).Value + BulkRestoreCoordinationTimeout;
timeoutTime = new TimeSpan(Math.Max(timeoutTime.Ticks, 0));
```

Reply:
on the other side of the coin for whatever reason the time out is < -1 then `Task.Delay` will throw an exception. May be this case will never happen not sure.

Reply:
Erick's right, it's a 5 min timeout. I'll account for the negative time.

Reply:
Fixed. I added a static method and some tests for that.